### ISSUE TITLE
DB model not accurate inside transactions

### ISSUE THREAD

**Issue by abelseniorly** [go-gorm/gorm#4117]
## GORM Playground Link

https://github.com/go-gorm/playground/pull/287

## Description

The DB model is not updated inside a transaction, and it reuses the model from previous db call (even if started using new session call).. Major bug while using transactions

**Comment by Xieyuschen** [comment]
I think it ensures thread safety.
Assume if a user share a `db *gorm.DB` in his project, he operates `db` in two go routine just like this:
```go
func main(){
	var db *gorm.DB	
       //routine 1
        go func() {
		db.Where("id=?",3).Find(&result)
	}()

       //routine 2
	go func() {
		db.Where("name=?",demo).Find(&result)
	}()
}
```
What is the result of the above code? Just a mistory for there are two goroutine use the same `db`. Assume your thought, `The DB model is not updated inside a transaction`. Even though it's transaction, but because they share `db` , so chainable method will affect the result. 
Like the codes above, it may make two query like:
```sql
- Case 1,the routine1 execute `Where("id=?",3)` first and query, then routine2 query
       -routine 1 sql:
Select * from result where id=3;
       -routine 2 sql:
Select * from result where id=3 and name = 'demo' ;
- Case 2,the routine2 execute Where("name=?",demo) and query, then routine2 query
       -routine 1 sql:
Select * from result where id=3 and name = 'demo' ;
       -routine 2 sql:
Select * from result where  name = 'demo' ;
```
So it will dangerous when it comes to concurrency, so to ensure thread safety, you a transaction should not update the db it use. Instead, it return a db which is updated.

**Comment by jinzhu** [comment]
this is expected behaviour, refer https://gorm.io/docs/method_chaining.html

**Comment by abelseniorly** [comment]
@Xieyuschen @jinzhu here I explicitely create a new Session with the function before starting the transaction itself... Why is the statement still cloned there?

**Comment by Xieyuschen** [comment]
There maybe many transactions in the same session. If you update it, it will seriously affect the query result it you do those transactions in different go routine. To avoid this, it won't do such.
I think it's designed for this, what about you?

**Comment by abelseniorly** [comment]
Isnt a new session with newDB: true supposed to empty the models and conditions and return a fresh new db connection? @jinzhu

**Comment by Xieyuschen** [comment]
It not return a new db connection, just update the statement of the `db`. DB connection is managed automatically and seems not exposed to user.

**Comment by abelseniorly** [comment]
I have updated the test in playground.... The `db.Session(&gorm.Session{NewDB: true})` is supposed to clear all the conditions... However when the transaction is begun, it is not cleared..

**Comment by Xieyuschen** [comment]
Oh let me have a try, I just read source code to learn about `session` and had no experiences.
